--- lib/hostip4.c.orig	2011-09-05 09:42:41.696922780 +0800
+++ lib/hostip4.c	2011-09-05 09:40:35.656389768 +0800
@@ -161,6 +161,12 @@
       sbufptr = sbuf;
     }
     hints.ai_flags = AI_CANONNAME;
+#ifdef HAVE_GSSAPI
+ if(conn->data->set.krb)
+     /* if krb is used, we (might) need the canonical host name */
+    hints.ai_flags |= AI_CANONNAME;
+#endif
+
     (void)Curl_getaddrinfo_ex(hostname, sbufptr, &hints, &ai);
 
 #elif defined(HAVE_GETHOSTBYNAME_R)
